src/base/http/connection.cpp

```diff
@@ -33,29 +34,122 @@
 #include <QTcpSocket>
 
 #include "base/logger.h"
-#include "irequesthandler.h"
 #include "requestparser.h"
-#include "responsegenerator.h"
 
 using namespace Http;
 
-Connection::Connection(QTcpSocket *socket, IRequestHandler *requestHandler, QObject *parent)
-    : QObject(parent)
-    , m_socket(socket)
-    , m_requestHandler(requestHandler)
+Connection::Connection(QTcpSocket *socket, QObject *parent)
+    : QObject {parent}
+    , m_socket {socket}
 {
     m_socket->setParent(this);
-    m_idleTimer.start();
-    connect(m_socket, &QTcpSocket::readyRead, this, &Connection::read);
+    m_receivedData.reserve(RequestParser::MAX_CONTENT_SIZE * 1.1);
+    m_statusHeaderBuffer.reserve(RequestParser::MAX_CONTENT_SIZE * 1.1);
+
+    connect(m_socket, &QAbstractSocket::readyRead, this, &Connection::read);
+    connect(m_socket, &QAbstractSocket::bytesWritten, this, &Connection::bytesWritten);
+    connect(m_socket, &QAbstractSocket::disconnected, this, &Connection::disconnected);
+}
+
+void Connection::sendStatus(const Http::ResponseStatus &status)
+{
+    Q_ASSERT(m_status == PendingStatus);
```

Comment:
m_status can be fully removed. I actually inserted it for testing purposes but then decided to keep, it can be fully removed. 

> The public interface of a class should either be designed in such a way as to exclude the possibility of incorrect use itself, 

IMO this point is not applicable here, the HTTP response is well defined and users of the `Connection` class should be aware of it and use it such.

> (although it is possible to some extent). So it would be nice to have at least a second behavior

Do you mean to provide a direct Connection::write function?

Reply:
>Do you mean to provide a direct Connection::write function?

No.
I mean the following:
If some action on an object of a class may be invalid depending on its current state, then the caller side must have a way to check this internal state (like in my example with array above).
>the HTTP response is well defined and users of the Connection class should be aware of it and use it such.

Looks like you're right. In this case, it will probably be enough to leave the assert, or even get rid of m_status (if it is not required for something else).